core/shared/src/main/scala/org/typelevel/log4cats/LocalLogContext.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2018 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats
+
+import cats.mtl.{Ask, Local}
+import cats.syntax.functor.*
+import cats.syntax.traverse.*
+import cats.{~>, Applicative, Show}
+import org.typelevel.otel4s.KindTransformer
+
+import scala.collection.immutable.ArraySeq
+
+sealed trait LocalLogContext[F[_]] {
+  private[log4cats] def currentLogContext: F[Map[String, String]]
+
+  private[log4cats] def withAddedContext[A](ctx: Map[String, String])(fa: F[A]): F[A]
+
+  private[log4cats] final def withAddedContext[A](ctx: (String, Show.Shown)*)(fa: F[A]): F[A] =
+    withAddedContext {
+      ctx.view.map { case (k, v) => k -> v.toString }.toMap
+    }(fa)
+
+  def withHighPriorityAskedContext(ask: Ask[F, Map[String, String]]): LocalLogContext[F]
+
+  def withLowPriorityAskedContext(ask: Ask[F, Map[String, String]]): LocalLogContext[F]
```

Comment:
these allow you to, for example, add trace and span IDs to the context. high/low priority is which wins if there's a key collision (hopefully never happens for span context, but could for other things)